### PR TITLE
[READY] New feature #13513 : allow plugins to add expresssion manager function

### PR DESCRIPTION
**Pull request by Shnoulle**
Must be tested ;) and surely improved

### CODE REVIEW

**Comment by Shnoulle** [comment]
Some question / remark : 

- If needed we can add more event here : https://github.com/LimeSurvey/LimeSurvey/pull/1081/files#diff-7f2b63ae2aaab1ffe6caffbaae39751cR259 ;)
- Move is_function to is_callable allow namespace (or not) https://github.com/LimeSurvey/LimeSurvey/pull/1081/files#diff-7f2b63ae2aaab1ffe6caffbaae39751cR1841
- For javascript : creation of package, maybe add an helper in PluginManager for this. With package it can be external script (CDN of [phpjs](http://locutus.io/php/) ;) )

**Comment by Shnoulle** [comment]
Sh… using an old version …

**Comment by Shnoulle** [comment]
Leave open until open another one

**Comment by Shnoulle** [comment]
Found my issue with travis :) fix it today

**Comment by Shnoulle** [comment]
The issue in travis are now [same than master](https://github.com/LimeSurvey/LimeSurvey/commit/787b55403672c3aff165f205e5fd44c6c50bdb78) :) 

Then not related of this commit :)

**Comment by Shnoulle** [comment]
Some comment now : 

Another solution can be

1. Add function to add/fix expression in em (like addPackageForExpressionManager and RegisterFunctions)
2. Plugin can directly use it at some point , but we need to be sure EM is started at this point (unsure it was started in beforeSurveyPage)
3. Core can use it to allow upload/function system with twig file

Maybe leave the event, but add a system for twig extensions : https://github.com/LimeSurvey/LimeSurvey/pull/1081/files#diff-a05244074e414b80bca5b52851afaeb3R56 , replacing setPathOfAlias by some twig call system, javascript package can use addPackages and adding it to the new package

**Comment by Shnoulle** [comment]
Test survey (currently on this branch) `master.sondages.pro/764563?newtest=Y&lang=en`

**Comment by Shnoulle** [comment]
Issue is travis : https://github.com/LimeSurvey/LimeSurvey/commit/c425c278b0c3096e9ae6dda00c2bc38747198b14 not related to this PR.

**Comment by olleharstedt** [comment]
That commit _fixes_ the issue. Please merge with master and push again.

But yeah, it's working.

**Comment by Shnoulle** [comment]
there are 2 new issue in codefactor : javascript function not used … but all EM js function are not used …

`'doHtmlList' is defined but never used.  (lines of code = 1)` and ` 'sayHello' is defined but never used.  (lines of code = 1) ` 

I don't know if we can ad comment to disable this part of codefactor

**Comment by Shnoulle** [comment]
Arg … surely merge issue … non time today …

**Comment by Shnoulle** [comment]
@olleharstedt : really fixed now . Did you find improvment ? I'm unsure about the event part.

1. current solution add `'packages'` for javascript, this package is added to public view : then allow javascript function
2. But an alternative can be : just adding function, the plugin MUST add it itself to package : in beforeControllerAction or in the vent itself.

Maybe 2 is best BUT with an helper in PluginManager : $this->addEmPackage + $this->addPathOfAlias ?

Something like this. The real update is https://github.com/LimeSurvey/LimeSurvey/pull/1081/files#diff-7f2b63ae2aaab1ffe6caffbaae39751cL1823  , after is more «Way to call»

@LouisGac : i think code can be used to allow function by twig : adding a core EM function : callTwigFile : function twig extension must : ell the function name, params etc … + the twig file to be used, the system call something like this (for php): 

````
callTwigFile($twigfile,$arg) {
    return $this->_twig->loadTemplate($twigfile.'.twig')->render(['arg'=>$arg]);
}
````

**Comment by Shnoulle** [comment]
Moving to develop, put in WIP time to test and fix

**Comment by Shnoulle** [comment]
I merge and review for last develop and dev team discussion

File: application/helpers/expressions/em_core_helper.php [hunk 1]
```diff
@@ -1849,7 +1877,7 @@ private function RDP_RunFunction($funcNameToken, $params)
                     switch ($argsPassed) {
                         case 0:
                             if (!$this->RDP_onlyparse) {
-                                $result = $funcName();
+                                $result = call_user_func($funcName);
```
Comment (olleharstedt) [inline application/helpers/expressions/em_core_helper.php]: Any difference, really?

File: application/config/packages.php [hunk 2]
```diff
@@ -304,8 +304,13 @@
         'depends' => array(
             'adminbasics',
         )
-        ),
-
+    ),
+    /* An empty package to be extended for EM (after core expressions) */
```
Comment (olleharstedt) [inline application/config/packages.php]: +1 for commenting the code.

File: plugins/Demo/demoAddEmFunction/assets/demoAddEmFunction.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+function sayHello(message) {
```
Comment (olleharstedt) [inline plugins/Demo/demoAddEmFunction/assets/demoAddEmFunction.js]: +1 for adding plugin demo!

File: application/helpers/expressions/em_core_helper.php (hunk 1 above)
Comment (Shnoulle) [reply application/helpers/expressions/em_core_helper.php]: Didn't test, but i dislike to use $funcName like this … and unsure it don't make difference with complete class …

File: plugins/Demo/demoAddEmFunction/assets/demoAddEmFunction.js (hunk 3 above)
Comment (Shnoulle) [reply plugins/Demo/demoAddEmFunction/assets/demoAddEmFunction.js]: Maybe adding javascript Class for function :/

**Comment by Shnoulle** [comment]
Last commit

- rename event to ExpressionManagerStart
- wrap plugin javascript function in Classe
- Add a demo survey
- fix javascript file can be not registred (expression-extend not registred)

**Comment by Shnoulle** [comment]
@ovillani : i think it totally add the needed feature :).

See the demo plugin : https://github.com/LimeSurvey/LimeSurvey/tree/ca0095f066bddf01f80723341406bf39a9485ce1/plugins/Demo/demoAddEmFunction

A php file for php function and a js file for javascript function.

I the included php file we can directly use 
```
<?php

function sayHello($message) {
    echo $message;
}
```
(like in js). But really : plugin must use Class name and nameSpace …

PS : i think plugin can replace current function (?) to be tested : is this needed ?
PS2 : to disallow this : reverse https://github.com/LimeSurvey/LimeSurvey/blob/30d797bd0502aad0d31737337c506286cac555ca/application/helpers/expressions/em_core_helper.php#L1973 (maybe need `array_merge_recursive`)

**Pull request merged**